sale_invoice_plan/models/sale.py

```diff
@@ -0,0 +1,223 @@
+    def action_confirm(self):
+        for rec in self:
+            if rec.use_invoice_plan and not rec.invoice_plan_ids:
+                raise UserError(
+                    _('Use Invoice Plan selected, but no plan created'))
+        return super().action_confirm()
+
+    @api.multi
+    def create_invoice_plan(self, num_installment, installment_date,
+                            interval, interval_type, advance):
+        self.ensure_one()
+        self.invoice_plan_ids.unlink()
+        invoice_plans = []
+        Decimal = self.env['decimal.precision']
+        prec = Decimal.precision_get('Product Unit of Measure')
+        percent = round(1.0 / num_installment * 100, prec)
+        percent_last = 100 - (percent * (num_installment-1))
+        # Advance
+        if advance:
+            vals = {'installment': 0, 'plan_date': installment_date,
+                    'invoice_type': 'advance', 'percent': 0.0}
+            invoice_plans.append((0, 0, vals))
+            installment_date = self._next_date(installment_date,
+                                               interval, interval_type)
+        # Normal
+        for i in range(num_installment):
+            this_installment = i+1
+            if num_installment == this_installment:
+                percent = percent_last
+            vals = {'installment': this_installment,
+                    'plan_date': installment_date,
+                    'invoice_type': 'installment',
+                    'percent': percent}
+            invoice_plans.append((0, 0, vals))
+            installment_date = self._next_date(installment_date,
+                                               interval, interval_type)
+        self.write({'invoice_plan_ids': invoice_plans})
+        return True
+
+    @api.multi
+    def remove_invoice_plan(self):
+        self.ensure_one()
+        self.invoice_plan_ids.unlink()
+        return True
+
+    @api.model
+    def _next_date(self, installment_date, interval, interval_type):
+        installment_date = fields.Date.from_string(installment_date)
+        if interval_type == 'month':
+            next_date = installment_date + relativedelta(months=+interval)
+        elif interval_type == 'year':
+            next_date = installment_date + relativedelta(years=+interval)
+        else:
+            next_date = installment_date + relativedelta(days=+interval)
+        next_date = fields.Date.to_string(next_date)
+        return next_date
+
+    @api.multi
+    def action_invoice_create(self, grouped=False, final=False):
+        inv_ids = super().action_invoice_create(grouped=grouped, final=final)
+        invoice_plan_id = self._context.get('invoice_plan_id')
+        if invoice_plan_id:
+            plan = self.env['sale.invoice.plan'].browse(invoice_plan_id)
+            invoices = self.env['account.invoice'].browse(inv_ids)
+            invoices.ensure_one()  # Expect 1 invoice for 1 invoice plan
+            plan._compute_new_invoice_quantity(invoices[0])
```

Comment:
IMHO in '_compute_new_invoice_quantity', you should do a write on invoice values.

Reply:
Sorry, I don't quite understand. You mean moving 
`plan.invoice_ids += invoice`
inside _compute_new_invoice_quantity()?

Reply:
No, maybe I'm at the wrong place to do the comment.

But here you are in an action (launched by a button I suppose).

that's why you have to use write() in '_compute_new_invoice_quantity' instead of doing 'line.quantity = plan_qty'

Reply:
OK, I will fix that. By the way, I also curious. In what situation we should not use record.field = value?

Reply:
Usually, such affectation should only be used in compute and inverse functions as the write is done after (with recomputes and so on). When it is launched through an action or,... you are not sure that a write will be done(and good recomputes on fields that depends on the one you've modified).